### PR TITLE
Entity details update

### PR DESCRIPTION
**Pull request by Lukewh**
Implements GUI part of https://github.com/CanonicalLtd/juju-design/issues/200

### CODE REVIEW

**Comment by Lukewh** [comment]
## QA
- Pull the branch
- Go to various bundles and charms
- Make sure the details pages look like the designs and still function properly

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7948/

**Comment by huwshimi** [comment]
Thanks a lot @Lukewh, lots of work done here! Here are my QA notes:

Charms:
- shadow missing from the deploy command input (although it's also missing from the search box above it etc.)
- in the header "Latest Revision" should be "Latest version" and there is no # before the number
- the "Learn more" link under "Embed this charm" is grey and bold (the design has it blue and normal)
- the input for selecting the embed coder is a different type of input (doesn't have the copy button etc.)
- the links inside the readme are grey instead of blue
- the Configuration section does not have space above it
- the Configuration value type e.g. "(string)" is bolder than the design

Bundles:
- there is a series in the header, but I can't remember if we should show that or not (it's not in the design)
- do we have a dscription for the bundles? If so it's missing.
- you can't click on the diagram to expand it
- the bundle config items that aren't expandable still have a hand pointer
- the lines between the config items are solid instead of dotted and the're is a little too much spacing between each item

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7954/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7955/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7956/

**Comment by Lukewh** [comment]
@huwshimi thanks for playing spot the difference - can you have another go and see if you can beat your previous score? Cheers!

**Comment by huwshimi** [comment]
Thanks @Lukewh! Only the tiniest change: "Latest **V**ersion" should be "Latest **v**ersion" in the header

File: jujugui/static/gui/src/app/components/entity-details/content/content.js [hunk 1]
```diff
@@ -636,33 +642,37 @@ EntityContent.defaultProps = {
 };
 
 EntityContent.propTypes = {
-  addNotification: PropTypes.func.isRequired,
-  apiUrl: PropTypes.string.isRequired,
-  changeState: PropTypes.func.isRequired,
-  entityModel: PropTypes.object.isRequired,
+  addNotification: React.PropTypes.func.isRequired,
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/entity-details/content/content.js]: These should all be `PropTypes...` ready for the React upgrade.

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/entity-details/content/content.js]: These should all be `PropTypes` ready for the React upgrade.

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js [hunk 2]
```diff
@@ -19,29 +19,67 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 'use strict';
 
 class EntityContentDiagram extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      url: this.props.getDiagramURL(this.props.id)
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: Because this is never used more than once on load I don't see why we couldn't require that a valid url be passed into the component. Dropping the requirement for `state` and for the `getDiagramURL` prop.

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js [hunk 3]
```diff
@@ -19,29 +19,67 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+    };
+  }
+
+  _handleClose() {
+    this.props.clearLightbox();
+  }
+
+  _handleExpand() {
+    console.log(this.props.title);
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: extra log

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js [hunk 4]
```diff
@@ -19,29 +19,67 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      url: this.props.getDiagramURL(this.props.id)
+    };
+  }
+
+  _handleClose() {
+    this.props.clearLightbox();
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: Can the lightbox close itself? Maybe it should handle its own closings?

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js [hunk 5]
```diff
@@ -19,29 +19,67 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
   }
 };
 
 EntityContentDiagram.propTypes = {
+  clearLightbox: PropTypes.func,
+  displayLightbox: PropTypes.func,
   getDiagramURL: PropTypes.func.isRequired,
   id: PropTypes.string.isRequired,
-  isRow: PropTypes.bool
+  isExpandable: PropTypes.bool,
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: out of curiosity, when would it not be?

File: jujugui/static/gui/src/app/components/entity-details/content/files/files.js [hunk 6]
```diff
@@ -27,12 +32,12 @@ class EntityFiles extends React.Component {
...
-    var isCollapsed = target.className.indexOf('collapsed') > 0;
-    target.className = classNames(
-      'entity-files__directory',
-      {'collapsed': !isCollapsed}
-    );
+    const target = e.currentTarget;
+    const controls = target.getAttribute('aria-controls');
+    const isExpanded = target.getAttribute('aria-expanded') === 'true';
+    this.setState({
+      [`${controls}-isExpanded`]: !isExpanded
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/files/files.js]: yay computed property names....sure ugly though ;)

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js [hunk 7]
```diff
@@ -19,29 +19,67 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 'use strict';
 
 class EntityContentDiagram extends React.Component {
+  constructor(props) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: Need to add tests for these additions.

File: jujugui/static/gui/src/app/components/entity-details/content/files/files.js [hunk 8]
```diff
@@ -137,21 +142,32 @@ class EntityFiles extends React.Component {
...
-            onClick={this._onDirectoryClick.bind(this)}>
-            {'/' + fileName}
-            <ul className="entity-files__listing">
+            title={`/${fileName}`}>
+            <button className="p-list-tree__toggle"
+              id={`/${fileName}-toggle`}
+              role="tab"
+              onClick={this._onDirectoryClick.bind(this)}
+              aria-controls={`/${fileName}`}
+              aria-expanded={`${isExpanded}`}>
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/files/files.js]: 👍

File: jujugui/static/gui/src/app/components/entity-details/content/files/files.js [hunk 9]
```diff
@@ -161,9 +177,9 @@ class EntityFiles extends React.Component {
 
     // Need to handle the root node separate from the recursive logic.
     // The loop that covers everything *but* the root node is above.
-    var markup = [];
+    let markup = [];
     Object.keys(filetree).forEach(file => {
-      markup.push(buildList.call(this, file, filetree[file]));
+      markup.push(buildList.bind(this)(file, filetree[file]));
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/files/files.js]: Why this change? Functionally the same but less performant...

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js [hunk 10]
```diff
@@ -283,6 +283,16 @@ const ComponentRenderersMixin = (superclass) => class extends superclass {
       document.getElementById('modal-gui-settings'));
   }
 
+  _displayLightbox(content, caption) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: docblock

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js [hunk 11]
```diff
@@ -296,6 +306,13 @@ const ComponentRenderersMixin = (superclass) => class extends superclass {
     ReactDOM.unmountComponentAtNode(
       document.getElementById('modal-gui-settings'));
   }
+
+  _clearLightbox() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: docblock

File: jujugui/static/gui/src/app/components/lightbox/lightbox.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
+/* Copyright (C) 2015 Canonical Ltd. */
+
+'use strict';
+
+class Lightbox extends React.Component {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/lightbox/lightbox.js]: could you add a comment here for what the component is/for

File: jujugui/static/gui/src/app/components/entity-details/content/content.js [hunk 13]
```diff
@@ -348,6 +336,28 @@ class EntityContent extends React.Component {
...
+  _generateDiagram(entityModel) {
+    if (entityModel.get('entityType') !== 'bundle') {
+      return;
+    }
+    const entity = entityModel.toEntity();
+    return <juju.components.EntityContentDiagram
+      clearLightbox={this.props.clearLightbox}
+      displayLightbox={this.props.displayLightbox}
+      getDiagramURL={this.props.getDiagramURL}
+      id={entityModel.get('id')}
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/entity-details/content/content.js]: this could become:

`diagramURL={props.getDiagramURL(entitymode.get('id'))}`

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js (hunk 4 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: You can't remove a component from within itself, so I followed the convention used for the modals etc.

File: jujugui/static/gui/src/app/components/entity-details/content/content.js (hunk 1 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/content.js]: Done

File: jujugui/static/gui/src/app/app.js [hunk 14]
```diff
@@ -1888,6 +1890,16 @@ YUI.add('juju-gui', function(Y) {
         document.getElementById('modal-gui-settings'));
     },
 
+    _displayLightbox: function(content, caption) {
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/app.js]: Done

File: jujugui/static/gui/src/app/app.js [hunk 15]
```diff
@@ -1904,6 +1916,12 @@ YUI.add('juju-gui', function(Y) {
         document.getElementById('modal-gui-settings'));
     },
 
+    _clearLightbox: function() {
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/app.js]: Done

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js (hunk 10 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: Done

Comment (Lukewh) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: Done

File: jujugui/static/gui/src/app/init/component-renderers-mixin.js (hunk 11 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: Done

Comment (Lukewh) [inline jujugui/static/gui/src/app/init/component-renderers-mixin.js]: Done

File: jujugui/static/gui/src/app/components/lightbox/lightbox.js (hunk 12 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/lightbox/lightbox.js]: Done

Comment (Lukewh) [inline jujugui/static/gui/src/app/components/lightbox/lightbox.js]: Done

File: jujugui/static/gui/src/app/components/entity-details/content/content.js [hunk 16]
```diff
@@ -348,6 +336,28 @@ class EntityContent extends React.Component {
   }
 
   /**
+  Generate the diagram markup for a bundle.
+
+  @method _generateDiagram
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/content.js]: Done

File: jujugui/static/gui/src/app/components/entity-details/content/content.js (hunk 13 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/content.js]: Done

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js (hunk 2 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: Done

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js (hunk 3 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: Done

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js (hunk 5 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: When in direct deploy

File: jujugui/static/gui/src/app/components/entity-details/content/files/files.js (hunk 9 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/files/files.js]: I think I removed it then rewrote it later... I've changed it back!

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js (hunk 4 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: This is never called or used, so I've removed it.

File: jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js (hunk 7 above)
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/entity-details/content/diagram/diagram.js]: Done

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7962/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7963/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7964/

**Comment by Lukewh** [comment]
Thanks @huwshimi and @hatch :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge
